dedupe/labeler.py

```diff
@@ -26,155 +26,107 @@
 
-    def fit_transform(self, pairs: TrainingExamples, y: LabelsLike) -> None:
-        self.fit(self.transform(pairs), y)
-
-    def pop(self) -> TrainingExample:
-        if not len(self.candidates):
-            raise IndexError("No more unlabeled examples to label")
-
-        target_uncertainty = self._bias()
-
-        probabilities = self.candidate_scores()
-
-        distance_to_target = numpy.abs(target_uncertainty - probabilities)
-        uncertain_index = distance_to_target.argmin()
-
-        self.distances = numpy.delete(self.distances, uncertain_index, axis=0)
-
-        uncertain_pair: TrainingExample = self.candidates.pop(uncertain_index)
-
-        return uncertain_pair
-
-    def _remove(self, index: int) -> None:
-        self.distances = numpy.delete(self.distances, index, axis=0)
-
-    def mark(self, pairs: TrainingExamples, y: LabelsLike) -> None:
-
-        self.y = numpy.concatenate([self.y, y])  # type: ignore[arg-type]
-        self.X = numpy.vstack([self.X, self.transform(pairs)])
-
-        self.fit(self.X, self.y)
+    @abstractmethod
+    def candidate_scores(self) -> numpy.typing.NDArray[numpy.float_]:
+        """For each of self.candidates, return our current guess [0,1] of if a match."""
 
-    def _bias(self) -> float:
-        positive: int = numpy.sum(self.y == 1)
-        n_examples = len(self.y)
+    @abstractmethod
+    def remove(self, index: int) -> None:
+        """Remove a pair from self.candidates."""
 
-        bias = 1 - (positive / n_examples if positive else 0)
 
-        # When we have just a few examples we are okay with getting
-        # examples where the model strongly believes the example is
-        # going to be positive or negative. As we get more examples,
-        # prefer to ask for labels of examples the model is more
-        # uncertain of.
-        uncertainty_weight = min(positive, n_examples - positive)
-        bias_weight = 10
+class MatchLearner(Learner):
+    def __init__(self, data_model: DataModel, candidates: TrainingExamples):
+        self.data_model = data_model
+        self._candidates = candidates.copy()
+        self._classifier = sklearn.linear_model.LogisticRegression()
+        self._distances = self._calc_distances(self.candidates)
+        self._fitted = False
 
-        weighted_bias = 0.5 * uncertainty_weight + bias * bias_weight
-        weighted_bias /= uncertainty_weight + bias_weight
+    def fit(self, pairs: TrainingExamples, y: LabelsLike) -> None:
+        self._classifier.fit(self._calc_distances(pairs), numpy.array(y))
 
-        return weighted_bias
+    def remove(self, index: int) -> None:
+        self._candidates.pop(index)
+        self._distances = numpy.delete(self._distances, index, axis=0)
 
     def candidate_scores(self) -> numpy.typing.NDArray[numpy.float_]:
-        scores: numpy.typing.NDArray[numpy.float_] = self.predict_proba(self.distances)[
-            :, 1
-        ].reshape(-1, 1)
+        if not self._fitted:
```

Comment:
if we are not fitted, then let's have the candidate scores return an array of "0.5" of length candidates

Reply:
or an error.

Reply:
I like ValueError more. Done.